components/de_amp/browser/de_amp_url_loader.cc

```diff
@@ -62,54 +64,72 @@ DeAmpURLLoader::~DeAmpURLLoader() = default;
 
 void DeAmpURLLoader::OnBodyReadable(MojoResult) {
   if (state_ == State::kSending) {
-    // The pipe becoming readable when kSending means all buffered body has
-    // already been sent.
-    ForwardBodyToClient();
+    if (bytes_remaining_in_buffer_ > 0) {
+      SendBufferedBodyToClient();
+    } else {
+      ForwardBodyToClient();
+    }
     return;
   }
   if (!CheckBufferedBody(kReadBufferSize)) {
     return;
   }
-
-  if (!MaybeRedirectToCanonicalLink()) {
+  const bool redirected = MaybeRedirectToCanonicalLink();
+  const bool found_amp_but_not_canonical_link = !redirected && found_amp_;
+
+  // If we were not redirected (navigation is cancelled) and we didn't find AMP,
+  // or if we did find AMP previously and we've already read more bytes than
+  // max, complete the load.
+  if ((!redirected && !found_amp_) ||
+      (found_amp_but_not_canonical_link && read_bytes_ >= kMaxBytesToCheck)) {
+    found_amp_ = false;  // reset
     CompleteLoading(std::move(buffered_body_));
```

Comment:
return is missed here
https://bravesoftware.slack.com/archives/C3T9S9WKD/p1660570912235569?thread_ts=1659383582.408519&cid=C3T9S9WKD